packages/pwa-kit-dev/src/utils/resolver-utils.js

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import path from 'path'
+import {getConfig} from '@salesforce/pwa-kit-runtime/utils/ssr-config'
+
+const EXTENSION_NAMESPACE = '@salesforce'
+const EXTENSION_PREFIX = 'extension'
+const NODE_MODULES = 'node_modules'
+const SDK_COMPONENT_MAP = {
+    'app/routes': '/ssr/universal/components/routes'
+}
+const INDEX_FILE = 'index'
+
+// Returns true/false indicating if the importPath resolves to a same named file as the sourcePath.
+// @private
+export const isSelfReference = (importPath, sourcePath) => {
+    const indexRegExp = new RegExp(`(/${INDEX_FILE})$`)
+
+    // Sanitize the input. Here we want to remove the file extension and index file if it exists.
+    sourcePath = sourcePath.split('.')[0]
+    sourcePath = sourcePath.split(path.sep).join('/')
+    sourcePath = sourcePath.replace(indexRegExp, '')
+
+    // Do the same for the import path even thought it's not common to use /index and file extensions in your module
+    // imports.
+    importPath = importPath.split('.')[0]
+    importPath = importPath.replace(indexRegExp, '')
+
+    return sourcePath.endsWith(importPath)
+}
+
+/**
+ * Normalize and expand the extension configuration array so that it is easier to process.
+ * @param {{String, Object}[]} extensions - The extensions configuration value as defined in the PWA-Kit config.
+ * @returns {Object[]} extensions - The extensions array in object form.
+ *
+ * @example
+ * const result = expand(["store-finder", ["account-pages", {singlePage: true}], './extensions/local-extension']);
+ * console.log(result)
+ * // [["@salesforce/extension-store-finder", {}], ["@salesforce/extension-account-pages", {singlePage: true}], ["/home/project/extensions/local-extension", {}]]
+ */
```

Comment:
Thank you for being thoughtful! This is that kind of function where it's really hard to understand without an example.